checker/src/org/checkerframework/checker/index/searchindex/SearchIndexAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,112 @@
+import org.checkerframework.framework.util.AnnotationBuilder;
+import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/**
+ * The Search Index Checker is used to help type the results of calls to the JDK's binary search
+ * methods. It is part of the Index Checker.
+ */
+public class SearchIndexAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    public final AnnotationMirror UNKNOWN, BOTTOM;
+
+    public SearchIndexAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        UNKNOWN = AnnotationUtils.fromClass(elements, SearchIndexUnknown.class);
+        BOTTOM = AnnotationUtils.fromClass(elements, SearchIndexBottom.class);
+        this.postInit();
+    }
+
+    /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant propagation and folding).
+     */
+    ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
+        return getTypeFactoryOfSubchecker(ValueChecker.class);
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return new LinkedHashSet<>(
+                Arrays.asList(
+                        SearchIndexFor.class,
+                        SearchIndexBottom.class,
+                        SearchIndexUnknown.class,
+                        NegativeIndexFor.class));
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(
+            MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+        return new SearchIndexQualifierHierarchy(factory);
+    }
+
+    private final class SearchIndexQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public SearchIndexQualifierHierarchy(
+                MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexUnknown.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexBottom.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexUnknown.class)) {
+                return false;
+            }
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexBottom.class)) {
+                return false;
+            }
+
+            // Each annotation is either NegativeIndexFor or SearchIndexFor.
+            List<String> superArrays = IndexUtil.getValueOfAnnotationWithStringArgument(superAnno);
+            List<String> subArrays = IndexUtil.getValueOfAnnotationWithStringArgument(subAnno);
+
+            // It's possible that either superArrays or subArrays is null in rare circumstances.
```

Comment:
I'm not sure if it's a framework bug or not, because I haven't investigated significantly. When the check is removed, the test I cited (all-systems/Issue1006.java) throws an NPE. Since it's a regression test, I assumed that that's the desired behavior, so I wrote the check. I can spend some time investigating if you think it's worthwhile, or I can file a CF issue with the information I have now.